kompile: add check for token sort productions

SchmErik opened the pull request:
Token sorts should only contain productions that have `token`,
`bracket`, or `function` labels or the RHS of the production is a macro
term. Any other label (including subsorts) are illegal.

The java backend is more forgiving when it comes to token sort
declarations. Avoid this check when kompile is running with the java
backend.

Fixes: #2119

cos commented on kore/src/main/scala/org/kframework/definition/transformers.scala:
```diff
@@ -66,7 +66,23 @@ abstract class ModuleTransformer extends (Module => Module) {
     case e: ModuleTransformerException => throw e
-    case e: Throwable => throw new ModuleTransformerException(name, e)
+    case e: Throwable => {
+      if (e.toString().contains("Can't mix cells")) {
```
Adjusting the exception based on the exception's description string is not a fix for this issue. Try to trace the exception to its source, see why the original message was wrong, and adjust there to throw the correct exception.

Also, this file contains generic transformer code, so it should not contain exception details regarding a particular pass.

daejunpark commented:
Close this PR exercise in the class.

Commit:
kompile: add check for token sort productions

Token sorts should only contain productions that have `token`,
`bracket`, or `function` labels or the RHS of the production is a macro
term. Any other label (including subsorts) are illegal.

Commit:
invalidConstructor check: update test results

Commit:
regression-new: add tests for token check

Commit:
kompile: do not run token checks when running java backend

The java backend is more forgiving when it comes to token sort
declarations. Avoid this check when kompile is running with the java
backend.

Commit:
pl-tutorial: add token labels to files kompiled using llvm backend

radumereuta commented on kernel/src/main/java/org/kframework/kompile/Kompile.java:
```diff
@@ -437,6 +438,8 @@ public void structuralChecks(scala.collection.Set<Module> modules, Module mainMo
 
         stream(modules).forEach(m -> stream(m.localSentences()).forEach(new CheckHOLE(errors, m)::check));
 
+        stream(modules).forEach(m -> stream(m.localSentences()).forEach(new CheckTokens(errors, m, isSymbolic && isKast)::check));
```
I think this check `isSymbolic && isKast` can be done before even entering this line.
So we would avoid this entire step instead of looking at each sentence and decide if we need to do something.

Commit:
CheckTokens: move conditional to structuralChecks

CheckTokens class does not need to evaluate this logic for every sentence.

Commit:
Merge branch 'master' into check-tokens-2119-v2

radumereuta commented on kernel/src/main/java/org/kframework/kompile/Kompile.java:
```diff
@@ -437,6 +438,10 @@ public void structuralChecks(scala.collection.Set<Module> modules, Module mainMo
 
         stream(modules).forEach(m -> stream(m.localSentences()).forEach(new CheckHOLE(errors, m)::check));
 
+        if (!(isSymbolic && isKast)) {
```
```suggestion
        if (!(isSymbolic && isKast)) { // if it's not the java backend
```

radumereuta committed:
Add comment.

radumereuta committed:
Merge branch 'master' into check-tokens-2119-v2

Commit:
Merge branch 'master' into check-tokens-2119-v2
